### DISCUSSION TITLE
Help querying a relationship to define an attribute on a model

### DISCUSSION THREAD

**Discussion by aaronsgithub** [sqlalchemy/sqlalchemy#10315] [Usage Questions] [answered]
I'm using sqlalchemy==2.0.19

In the below, `ProfileImage` has an index which allows only one row to have the `current` column set to `True`.

I'm looking for a way to create an attribute on `User` which retrieves this row as a `ProfileImage` instance, that is: where `id==ProfileImage.user_id` and `ProfileImage.current==True`.

I'd like to understand the various ways of doing this, and their tradeoffs.

Please correct me if I'm wrong, but I believe there are two approaches:
- column_property
- hybrid_property

And then I would need to consider whether there is an enclosing query providing a join, or if I'm doing a correlated subquery.

I'm also using `asyncpg` and so I'd also like to understand how the relationship loading strategy should be factored into this.

I've being staring at the docs trying to make sense of this but I'm a little stuck.  So if anyone is able to help it would be much appreciated.

```python

# myapp.auth.models

if TYPE_CHECKING:
    from myapp.files.models import ProfileImage

class User(Base):
    __tablename__ = "auth_user"

    id: Mapped[int] = mapped_column(primary_key=True)
    ...
    profile_images: Mapped[list["ProfileImage"]] = relationship()

    # we want an attribute or property which gets only the current ProfileImage for the user.
    profile_image = <use column_property or hybrid_property???>
    

# ------------------

# myapp.files.models
from sqlalchemy_file import FileField # github.com/jowilf/sqlalchemy-file

class ProfileImage(Base, CreatedUpdatedMixin):
    __tablename__ = "files_profile_image"
    
    id: Mapped[int] = mapped_column(primary_key=True)
    file: Mapped[UploadFile | None] = mapped_column(FileField, nullable=False)
    user_id: Mapped[int] = mapped_column(ForeignKey("auth_user.id"))

    # https://stackoverflow.com/questions/28166915/postgresql-constraint-only-one-row-can-have-flag-set
    current: Mapped[bool] = mapped_column(default=False)
    

    __table_args__ = (sa.Index("profile_image_current", current, unique=True, postgresql_where=(current == True)),)

```

Edit

I just realised, I probably would not want profile_image to be a Mapped attribute on User. I've edited that out.

**Comment by zzzeek** [comment, marked as answer]
both hybrid and column_property could be used here, although for this use case column_property is usually better as you can get your "profile_image" value without loading the whole related table of profile images.

you want to look at [my answer](https://github.com/sqlalchemy/sqlalchemy/discussions/10205#discussioncomment-6672054) in #10205 which illustrates the standard correlated subquery approach to this problem.    In fact I think the SELECT here would be simpler since you have a `current` column, so you don't need to use an aggregate function.

you probably need to attach this attribute where you define `ProfileImage`, assuming `myapp.files.models` depends on `myapp.auth.models` (or otherwise it can be the other way around, but one has to import the other here somehow to make `column_property()`):

```py

# myapp.files.models

class ProfileImage(Base, CreatedUpdatedMixin):
    __tablename__ = "files_profile_image"
   # ...

from myapp.auth.models import User
    
User.profile_image = column_property(
        select(Profile.file).where(Profile.user_id == User.id, ProfileImage.current == true).scalar_subquery()
    )
```

**Reply by CaselIT** [reply]
don't you need to also call `correlate_expect(User)`?
